notify/sns/sns.go

```diff
@@ -0,0 +1,200 @@
+	}
+
+	client := sns.New(sess, &aws.Config{Credentials: creds})
+	publishInput := &sns.PublishInput{}
+
+	if n.conf.TopicARN != "" {
+		publishInput.SetTopicArn(tmpl(n.conf.TopicARN))
+		messageToSend, isTrunc, err := validateAndTruncateMessage(tmpl(n.conf.Message))
+		if err != nil {
+			return false, err
+		}
+		if isTrunc {
+			n.conf.Attributes["truncated"] = "true"
+		}
+
+		if n.isFifo == nil {
+			checkFifo, err := checkTopicFifoAttribute(client, n.conf.TopicARN)
+			if err != nil {
+				if e, ok := err.(awserr.RequestFailure); ok {
+					return n.retrier.Check(e.StatusCode(), strings.NewReader(e.Message()))
+				} else {
+					return true, err
+				}
+			}
+			n.isFifo = &checkFifo
+		}
+		if *n.isFifo {
+			// Deduplication key and Message Group ID are only added if it's a FIFO SNS Topic.
+			key, err := notify.ExtractGroupKey(ctx)
+			if err != nil {
+				return false, err
+			}
+			publishInput.SetMessageDeduplicationId(key.Hash())
+			publishInput.SetMessageGroupId(key.Hash())
+		}
+
+		publishInput.SetMessage(messageToSend)
+	}
+	if n.conf.PhoneNumber != "" {
+		publishInput.SetPhoneNumber(tmpl(n.conf.PhoneNumber))
+		// If SMS message is over 1600 chars, SNS will reject the message.
+		_, isTruncated := notify.Truncate(tmpl(n.conf.Message), 1600)
+		if isTruncated {
+			return false, fmt.Errorf("SMS message exeeds length of 1600 charactors")
+		} else {
+			publishInput.SetMessage(tmpl(n.conf.Message))
+		}
+	}
+	if n.conf.TargetARN != "" {
+		publishInput.SetTargetArn(tmpl(n.conf.TargetARN))
+		messageToSend, isTrunc, err := validateAndTruncateMessage(tmpl(n.conf.Message))
+		if err != nil {
+			return false, err
+		}
+		if isTrunc {
+			n.conf.Attributes["truncated"] = "true"
+		}
+		publishInput.SetMessage(messageToSend)
+	}
+
+	if len(n.conf.Attributes) > 0 {
+		attributes := map[string]*sns.MessageAttributeValue{}
+		for k, v := range n.conf.Attributes {
+			attributes[tmpl(k)] = &sns.MessageAttributeValue{DataType: aws.String("String"), StringValue: aws.String(tmpl(v))}
+		}
+		publishInput.SetMessageAttributes(attributes)
+	}
+
+	if n.conf.Subject != "" {
+		publishInput.SetSubject(tmpl(n.conf.Subject))
+	}
+
+	publishOutput, err := client.Publish(publishInput)
+	if err != nil {
+		if e, ok := err.(awserr.RequestFailure); ok {
+			return n.retrier.Check(e.StatusCode(), strings.NewReader(e.Message()))
+		} else {
+			return true, err
+		}
+	}
+
+	err = n.logger.Log(publishOutput.String())
+	if err != nil {
+		return false, err
+	}
+
+	return false, nil
+}
+
+func checkTopicFifoAttribute(client *sns.SNS, topicARN string) (bool, error) {
+	topicAttributes, err := client.GetTopicAttributes(&sns.GetTopicAttributesInput{TopicArn: aws.String(topicARN)})
+	if err != nil {
+		return false, err
+	}
+	ta := topicAttributes.Attributes["FifoTopic"]
+	if ta != nil && *ta == "true" {
+		return true, nil
+	}
+	return false, nil
+}
+
+func validateAndTruncateMessage(message string) (string, bool, error) {
+	if utf8.ValidString(message) {
```

Comment:
[nit] You can reverse the logic using a guard here:

```suggestion
	if !utf8.ValidString(message) {
	    return "", false, fmt.Errorf("non utf8 encoded message string")
	}
```